src/containers/Application/ApplicationPage.tsx

```diff
@@ -2,32 +2,96 @@ import React, { useEffect, useState } from 'react'
 import { FullStructure, ResponsesByCode, ElementStateNEW } from '../../utils/types'
 import useGetFullApplicationStructure from '../../utils/hooks/useGetFullApplicationStructure'
 import { ApplicationStatus } from '../../utils/generated/graphql'
-import { useApplicationState } from '../../contexts/ApplicationState'
 import { useUserState } from '../../contexts/UserState'
 import { useRouter } from '../../utils/hooks/useRouter'
-import { Loading, NoMatch } from '../../components'
+import { Loading } from '../../components'
 import strings from '../../utils/constants'
-import messages from '../../utils/messages'
 import { Button, Grid, Header, Message, Segment, Sticky } from 'semantic-ui-react'
 import { PageElements } from '../../components/Application'
+import { useFormElementUpdateTracker } from '../../contexts/FormElementUpdateTrackerState'
 
 interface ApplicationProps {
   structure: FullStructure
   responses?: ResponsesByCode
 }
+// TODO move to shared types if used elsewhere
+type SectionAndPage = { sectionCode: string; pageName: string } | null
+
+interface MethodToCallOnRevalidation {
+  (firstInvalidPage: SectionAndPage): void
+}
+
+const getFirstInvalidPage: (
+  fullStructure: FullStructure | undefined
+) => { sectionCode: string; pageName: string } | null = (fullStructure) => {
+  // return { sectionCode: 'S1', pageName: 'Page 2' }
+  return null
+}
+
+interface RevalidationState {
+  methodToCallOnRevalidation: MethodToCallOnRevalidation | null
+  shouldProcessValidation: boolean
+  lastRevalidationRequest: number
+}
 
 const ApplicationPage: React.FC<ApplicationProps> = ({ structure }) => {
-  const [strictSectionPage, setStrictSectionPage] = useState({ section: null, page: null })
-  const { error, isLoading, fullStructure, responsesByCode } = useGetFullApplicationStructure({
-    structure,
-  })
   const {
     userState: { currentUser },
   } = useUserState()
   const { push, query } = useRouter()
 
+  const {
+    state: { isLastElementUpdateProcessed, elementUpdatedTimestamp },
+  } = useFormElementUpdateTracker()
+
+  const [strictSectionPage, setStrictSectionPage] = useState<SectionAndPage>(null)
+  const [revalidationState, setRevalidationState] = useState<RevalidationState>({
+    methodToCallOnRevalidation: null,
+    shouldProcessValidation: false,
+    lastRevalidationRequest: Date.now(),
+  })
+
+  const shouldRevalidate = isLastElementUpdateProcessed && revalidationState.shouldProcessValidation
+  const revalidateAfterTimestamp = shouldRevalidate ? elementUpdatedTimestamp : 0
+
+  const { error, fullStructure } = useGetFullApplicationStructure({
+    structure,
+    shouldRevalidate,
+    revalidateAfterTimestamp,
+  })
+
   const currentSection = query.sectionCode
-  const currentPage = `Page ${query.page}`
+  const currentPage = `Page${query.page}`
```

Comment:
I suggest spreading the param from URL here `{ sectionCode, page }` and just using this page to pass to other function - although the page is set to string I would cast to number to match to the one in the `FullStructure` on the `pages` value object `Page.number`.

Reply:
Hmm, I thought the idea of having page identifier in the key would be for easy of access ?

Otherwise have to wrap every iteration of search in Object.values. i.e. check out `getCurrentPageElements` file.

Btw, what's the difference between page key and seciton key (section key is a code but displayed through details and title) and look like page here is the page name for display